packages/2d/src/components/SVG.ts

```diff
@@ -0,0 +1,560 @@
+      size,
+      nodes,
+    };
+    SVG.svgNodesPool[svg] = builder;
+    return builder;
+  }
+
+  private static getElementTransformation(
+    element: SVGGraphicsElement,
+    parentTransform: DOMMatrix,
+  ) {
+    const transform = element.transform.baseVal.consolidate();
+    const transformMatrix = (
+      transform ? parentTransform.multiply(transform.matrix) : parentTransform
+    ).translate(
+      SVG.parseNumberAttribute(element, 'x'),
+      SVG.parseNumberAttribute(element, 'y'),
+    );
+    return transformMatrix;
+  }
+
+  private static getElementStyle(
+    element: SVGGraphicsElement,
+    inheritedStyle: ShapeProps,
+  ): ShapeProps {
+    return {
+      fill: element.getAttribute('fill') ?? inheritedStyle.fill,
+      stroke: element.getAttribute('stroke') ?? inheritedStyle.stroke,
+      lineWidth: element.hasAttribute('stroke-width')
+        ? parseFloat(element.getAttribute('stroke-width')!)
+        : inheritedStyle.lineWidth,
+      layout: false,
+    };
+  }
+
+  protected static getMatrixTransformation(transform: DOMMatrix): ShapeProps {
+    const position = {
+      x: transform.m41,
+      y: transform.m42,
+    };
+    const rotation = (Math.atan2(transform.m12, transform.m11) * 180) / Math.PI;
+    const scale = {
+      x: Vector2.magnitude(transform.m11, transform.m12),
+      y: Vector2.magnitude(transform.m21, transform.m22),
+    };
+    const determinant =
+      transform.m11 * transform.m22 - transform.m12 * transform.m21;
+    if (determinant < 0) {
+      if (transform.m11 < transform.m22) scale.x = -scale.x;
+      else scale.y = -scale.y;
+    }
+    return {
+      position,
+      rotation,
+      scale,
+    };
+  }
+
+  private static *extractGroupNodes(
+    element: SVGElement,
+    svgRoot: Element,
+    parentTransform: DOMMatrix,
+    inheritedStyle: ShapeProps,
+  ): Generator<SVGShapeData> {
+    for (const child of element.children) {
+      if (!(child instanceof SVGGraphicsElement)) continue;
+
+      yield* this.extractElementNodes(
+        child,
+        svgRoot,
+        parentTransform,
+        inheritedStyle,
+      );
+    }
+  }
+
+  private static parseNumberAttribute(
+    element: SVGElement,
+    name: string,
+  ): number {
+    return parseFloat(element.getAttribute(name) ?? '0');
+  }
+
+  private static *extractElementNodes(
+    child: SVGGraphicsElement,
+    svgRoot: Element,
+    parentTransform: DOMMatrix,
+    inheritedStyle: ShapeProps,
+  ): Generator<SVGShapeData> {
+    const transformMatrix = SVG.getElementTransformation(
+      child,
+      parentTransform,
+    );
+    const style = SVG.getElementStyle(child, inheritedStyle);
+    const id = child.id ?? '';
+    if (child.tagName === 'g') {
+      yield* SVG.extractGroupNodes(child, svgRoot, transformMatrix, style);
+    } else if (child.tagName === 'use') {
+      const hrefElement = svgRoot.querySelector(
+        (child as SVGUseElement).href.baseVal,
+      )!;
```

Comment:
Generally, I dislike the `!` operator to unnull something, especially because you effectively check for nulls on the next line anyways.